File: Leetcode/easy/middleNodeTortoiseAndHare.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Time: O(N)
        # Space: O(1)

        tortoise = head
        hare = head

        while hare and hare.next:
            tortoise = tortoise.next
            hare = hare.next.next

        return tortoise

'''

Runtime
38 ms
Beats
77.45%
Memory
13.9 MB
Beats
11.60%

'''

File: Leetcode/easy/middleNodeTortoiseAndHare.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        // Time: O(N)
        // Space: O(1)
        
        ListNode tortoise = head, hare = head;

        while (hare != null && hare.next != null) {
            tortoise = tortoise.next;
            hare = hare.next.next;
        }

        return tortoise;
    }
}

/*

Runtime
0 ms
Beats
100%
Memory
39.8 MB
Beats
89.65%

 */
